salt/transport/ipc.py

```diff
@@ -0,0 +1,305 @@
+import logging
+import socket
+import msgpack
+
+# Import Tornado libs
+import tornado
+import tornado.gen
+import tornado.netutil
+from tornado.ioloop import IOLoop
+from tornado.iostream import IOStream
+
+# Import Salt libs
+import salt.transport.client
+import salt.transport.frame
+
+log = logging.getLogger(__name__)
+
+
+class IPCServer(object):
+    '''
+    A Tornado IPC server very similar to Tornado's TCPServer class
+    but using either UNIX domain sockets or TCP sockets
+    '''
+    def __init__(self, opts, io_loop=None, stream_handler=None):
+        '''
+        Create a new Tornado IPC server
+
+        :param IOLoop io_loop: A Tornado ioloop to handle scheduling
+        :param func stream_handler: A function to customize handling of an
+                                    incoming stream.
+        '''
+        self.opts = opts
+        self._started = False
+        self.stream_handler = stream_handler
+
+        # Placeholders for attributes to be populated by method calls
+        self.stream = None
+        self.sock = None
+        self.io_loop = None
+
+    def start(self, socket_path):
+        '''
+        Perform the work necessary to start up a Tornado IPC server
+
+        Blocks until socket is established
+
+        :param str socket_path: Path on the filesystem for the socket to bind to.
+                                This socket does not need to exist prior to calling
+                                this method, but parent directories should.
+        '''
+        # Start up the ioloop
+        log.trace('IPCServer: binding to socket: {0}'.format(socket_path))
+        self.sock = tornado.netutil.bind_unix_socket(socket_path)
+
+        if self.io_loop is None:
+            self.io_loop = IOLoop.current()
+
+        tornado.netutil.add_accept_handler(
+            self.sock,
+            self.handle_connection,
+            io_loop=self.io_loop,
+        )
+        self._started = True
+
+    @tornado.gen.coroutine
+    def handle_stream(self, stream):
+        '''
+        Override this to handle the streams as they arrive
+
+        :param IOStream stream: An IOStream for processing
+
+        See http://tornado.readthedocs.org/en/latest/iostream.html#tornado.iostream.IOStream
+        for additional details.
+        '''
+        while True:
+            framed_msg_len = yield stream.read_until(' ')
+            framed_msg_raw = yield stream.read_bytes(int(framed_msg_len.strip()))
+            framed_msg = msgpack.loads(framed_msg_raw)
+            body = framed_msg['body']
+            self.io_loop.spawn_callback(self.stream_handler, body)
+
+    @tornado.gen.coroutine
+    def handle_connection(self, connection, address):
+        log.trace('IPCServer: Handling connection to address: {0}'.format(address))
+        try:
+            stream = IOStream(connection,
+                              io_loop=self.io_loop,
+                              )
+            yield self.handle_stream(stream)
```

Comment:
IIRC this means you'll do exactly one stream at a time. I think you want to spawn the callback, so you can have multiple in flight at once. (which would mean this method doesn't need to be a coroutine)